client/java/src/main/java/io/openlineage/client/utils/DatasetIdentifierUtils.java

```diff
@@ -5,88 +5,94 @@
-    if (name.chars().filter(x -> x == File.separatorChar).count() == 1 && name.startsWith("/")) {
-      return name.substring(1);
-    }
-    return name;
+  public static DatasetIdentifier fromComponents(URI namespace, String name, String defaultScheme) {
+    String normalizedNamespace = normalizeNamespace(namespace, defaultScheme);
+    String normalizedName = normalizeName(name);
+    return new DatasetIdentifier(normalizedName, normalizedNamespace);
   }
 
-  private static String removeLastSlash(String name) {
-    if (!name.isEmpty() && name.charAt(name.length() - 1) == File.separatorChar) {
-      return name.substring(0, name.length() - 1);
+  /* Convert namespace from URI to normalized string.
+   * Final slashes are removed, `/path` and `file:///path` are converted to `file:/path`
+   */
+  public static String normalizeNamespace(URI namespaceUri, String defaultScheme) {
+    String scheme = StringUtils.defaultIfEmpty(namespaceUri.getScheme(), defaultScheme);
+
+    String namespace;
+    try {
+      // Add missing schema, drop query and fragment
+      namespace =
+          new URI(scheme, namespaceUri.getAuthority(), namespaceUri.getPath(), null, null)
+              .toString();
+    } catch (URISyntaxException e) {
+      namespace = namespaceUri.toString();
     }
-    return name;
+
+    return Optional.of(namespace)
+        .map(s -> s.replace(AUTHORITY_SEPARATOR + SEPARATOR, SEPARATOR))
+        .map(s -> removeLastSlash(s))
+        .filter(s -> !s.equals(scheme + SCHEME_SEPARATOR))
+        .orElse(scheme);
   }
 
-  /**
-   * Copied implementation of `isAbsoluteAndSchemeAuthorityNull` method in Path class within hadoop
-   * common package. We don't want to add 4MB dependency, however we need to have a method that
-   * checks if a path is absolute in that way.
-   *
-   * @see
-   *     <a=href"https://github.com/apache/hadoop/blob/release-3.3.2-RC0/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/Path.java#L376">original
-   *     method</a>
+  /* Convert name to normalized string.
+   * Final slashes are removed, non-nested `/path` is converted to `path`, empty name is converted to `/`
    */
-  private static boolean isAbsoluteAndSchemeAuthorityNull(URI uri) {
-    if (uri.getScheme() != null || uri.getAuthority() != null) {
-      return false;
+  public static String normalizeName(String name) {
+    String normalizedName = removeFirstSlashIfSingleSlashInString(removeLastSlash(name));
+    if (normalizedName.isEmpty()) {
+      normalizedName = SEPARATOR;
     }
+    return normalizedName;
+  }
 
-    boolean hasWindowsDrive = (WINDOWS && HAS_DRIVE_LETTER_SPECIFIER.matcher(uri.getPath()).find());
-
-    int startPositionWithoutWindowsDrive = 0;
-    if (hasWindowsDrive) {
-      startPositionWithoutWindowsDrive = (uri.getPath().charAt(0) == SEPARATOR_CHAR ? 3 : 2);
+  private static String removeFirstSlashIfSingleSlashInString(String name) {
+    // But why?
```

Comment:
Could someone give an explanation why `/some/path` should start with `/` but `some-path` should not? I can't find any reference for this.

Reply:
What happens when you remove this? 
I think the answer can be found in `PathUtilsTest`: 
```
 di = PathUtils.fromURI(new URI("s3://data-bucket/path"), FILE);
 assertThat(di.getName()).isEqualTo("path");
 assertThat(di.getNamespace()).isEqualTo("s3://data-bucket");
```

Reply:
@dolfinus pls let me know if you want to do something about it within this PR?

Reply:
Let me rephrase my question.

Currently path like `s3://bucket/path` is converted to `namespace=s3://bucket, name=path`, but `s3://bucket/some/path` is converted to `namespace=s3://bucket, name=/some/path`. Is this required?

Reply:
> @dolfinus pls let me know if you want to do something about it within this PR?

Not now, probably in a separated PR.

Reply:
> Currently path like s3://bucket/path is converted to namespace=s3://bucket, name=path, but s3://bucket/some/path is converted to namespace=s3://bucket, name=/some/path. Is this required?

I think the leading slash should always be removed, unless it's the only character in name (f.e. `s3://bucket/` should be converted to namespace=s3://bucket and name=/)

Reply:
> I think the leading slash should always be removed

It is removed, but in some cases starting slash is removed too

Reply:
Is this applied to s3 only?

Reply:
Yes, that's what i meant - the starting / leading slash, the one at the beginning. Not sure why it's NOT removed sometimes, imho it should always be removed.

Reply:
```
 DatasetIdentifier di = PathUtils.fromURI(new URI("file:///home/test"), null);
    assertThat(di.getName()).isEqualTo("/home/test");
    assertThat(di.getNamespace()).isEqualTo(FILE);
```
If a name is path, then it makes sense for path to start with slash. The same applies for HDFS location:
```
di = PathUtils.fromPath(new Path("hdfs://namenode:8020/home/test"));
    assertThat(di.getName()).isEqualTo(HOME_TEST);
    assertThat(di.getNamespace()).isEqualTo("hdfs://namenode:8020");
```

I think the problem is that single method is used for multiple contexts (same as it used to be within jdbc url parsing). The logic may have sense for local files and hdfs, but makes less sense for object storages. Unfortunately, everything if implemented currently within the same code-flow.

Reply:
Do you mean that there should be multiple different classes implementing dataset namespace+name normalization, instead of one `DatasetIdentifierUtils`?

Reply:
I think yes - per-scheme one would be great? Similar to your contribution in https://github.com/OpenLineage/OpenLineage/pull/2773 :)

Reply:
Hm, ok, I'll try to implement this

Reply:
Done. Updated PR title and description to track the changes.